cdc/kv/client.go

```diff
@@ -1163,6 +1187,89 @@ func (s *eventFeedSession) singleEventFeed(
 	}
 }
 
+const scanLockLimit = 1024
+
+func (s *eventFeedSession) resolveLock(ctx context.Context, regionID uint64, maxVersion uint64) error {
+	// for scan lock request, we must return all locks even if they are generated
+	// by the same transaction. because gc worker need to make sure all locks have been
+	// cleaned.
+	req := tikvrpc.NewRequest(tikvrpc.CmdScanLock, &kvrpcpb.ScanLockRequest{
+		MaxVersion: maxVersion,
+		Limit:      scanLockLimit,
+	})
+
+	bo := tikv.NewBackoffer(ctx, tikv.GcResolveLockMaxBackoff)
+	var loc *tikv.KeyLocation
+	var key []byte
+	flushRegion := func() error {
+		var err error
+		loc, err = s.kvStorage.GetRegionCache().LocateRegionByID(bo, regionID)
+		if err != nil {
+			return err
+		}
+		key = loc.StartKey
+		return nil
+	}
+	if err := flushRegion(); err != nil {
+		return errors.Trace(err)
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+		req.ScanLock().StartKey = key
+		resp, err := s.kvStorage.SendReq(bo, req, loc.Region, tikv.ReadTimeoutMedium)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		regionErr, err := resp.GetRegionError()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if regionErr != nil {
+			err = bo.Backoff(tikv.BoRegionMiss, errors.New(regionErr.String()))
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if err := flushRegion(); err != nil {
+				return errors.Trace(err)
+			}
+			continue
+		}
+		if resp.Resp == nil {
+			return errors.Trace(tikv.ErrBodyMissing)
+		}
+		locksResp := resp.Resp.(*kvrpcpb.ScanLockResponse)
+		if locksResp.GetError() != nil {
+			return errors.Errorf("unexpected scanlock error: %s", locksResp)
+		}
+		locksInfo := locksResp.GetLocks()
+		locks := make([]*tikv.Lock, len(locksInfo))
+		for i := range locksInfo {
+			locks[i] = tikv.NewLock(locksInfo[i])
+		}
+
+		_, _, err1 := s.kvStorage.GetLockResolver().ResolveLocks(bo, 0, locks)
+		if err1 != nil {
+			return errors.Trace(err1)
+		}
+		if len(locks) < scanLockLimit {
+			key = loc.EndKey
+		} else {
+			key = locks[len(locks)-1].Key
+		}
+
+		if len(key) == 0 || bytes.Compare(key, loc.EndKey) >= 0 {
```

Comment:
```suggestion
		if len(key) == 0 || (len(loc.EndKey) != 0 && bytes.Compare(key, loc.EndKey) >= 0) {
```

`loc.EndKey` can be "", which means it's the last region. In this case if `key` is not "", it should continue scanning. You should check `loc.EndKey` like how TiDB did:

```go
if len(key) == 0 || (len(endKey) != 0 && bytes.Compare(key, endKey) >= 0) {
	break
}
```